src/frontend/api.ts

```diff
@@ -143,9 +152,113 @@ export const SERVER_START_TIMEOUT = 30000;
 
 const pixelRatio = PixelRatio.get();
 
-interface ApiParam {
-  baseUrl: string;
-  timeout?: number;
+function createRequestClient({
+  baseUrl,
+  onReady,
+}: {
+  baseUrl?: string;
+  onReady?: () => Promise<void>;
+} = {}) {
+  const req = ky.extend({
+    prefixUrl: baseUrl,
+    timeout: false,
+    headers: {
+      "cache-control": "no-cache",
+      pragma: "no-cache",
+    },
+  });
+
+  return {
+    get: async (url: string) => {
+      if (onReady) await onReady();
+      return await req.get(url).json();
+    },
+    del: async (url: string) => {
+      if (onReady) await onReady();
+      return await req.delete(url).json();
+    },
+    put: async (url: string, data: any) => {
+      if (onReady) await onReady();
+      return await req.put(url, { json: data }).json();
+    },
+    post: async (url: string, data: any) => {
+      if (onReady) await onReady();
+      return await req.post(url, { json: data }).json();
+    },
+  };
+}
+
+// TODO: Incorporate server status and making sure it's ready for requests?
+function createMapServerApi() {
+  let mapServerPort: number | undefined;
+  let client: ReturnType<typeof createRequestClient> | undefined;
+
+  nodejs.channel.addListener(
+    "map-server::start",
+    (payload: { port: number }) => {
+      if (mapServerPort !== payload.port) {
+        mapServerPort = payload.port;
+        client = createMapServerClient(mapServerPort);
+      }
+    }
+  );
+
+  function createMapServerClient(port: number) {
+    return createRequestClient({ baseUrl: getBaseUrl(port) });
+  }
+
+  function guaranteeClient() {
+    if (!mapServerPort)
+      throw new Error(
+        "Map server client cannot be used because port is unknown"
+      );
+
+    if (!client) {
+      client = createMapServerClient(mapServerPort);
+    }
+
+    return client;
+  }
```

Comment:
Haven't completely processed this, but what will happen if the front-end requests something from the map server before the `map-server::start` event has fired? I think it's quite likely the user will start up in the map screen and the front end would immediately make requests (similarly on a `resume` when returning to the app). I think we might need a `await mapServer.ready()` type pattern with a status listener?

Reply:
> it hasn't been an issue but i think that's because of luck and how the frontend waits for the app server before rendering the map

correction, it wasn't an issue because i was triggering api calls manually as opposed to some effect calling them on startup. so yeah, you're right that there's an issue with that in its current state